In this  we will going to clean data of students  raw data  

In [ ]:
!pip install numpy

In [13]:
import numpy as np
import pandas as pd
data=pd.read_csv("messy_students_1000.csv")
d_copy=data.copy()
d_copy

,Student_ID,Name,Age,Gender,Attendance,Python,DSA,DBMS,Maths,OS,Projects,Internships,Backlogs,CGPA,Placed,Admission_Date
0,1528,Arjun Mishra,19.0,Female,45.0,74.0,81.0,100.0,73.0,60.0,6.0,1.0,2.0,7.73,No,2024-12-09
1,1360,Tanya Gupta,20.0,Male,86.4,49.0,86.0,86.0,86.0,100.0,6.0,1.0,2.0,8.71,Yes,2024-06-24
2,1448,Ishita Agarwal,20.0,Female,77.6,69.0,76.0,55.0,93.0,84.0,1.0,0.0,3.0,7.65,Yes,2024-09-20
3,1032,Vivek Singh,18.0,Female,74.3,62.0,76.0,68.0,-8,41.0,0.0,1.0,2.0,8.86,No,2023-08-01
4,1622,Yash Mishra,19.0,Female,91.2,66.0,67.0,77.0,76.0,98.0,5.0,2.0,0.0,8.45,No,13/03/2025
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1020,1701,Simran Yadav,20.0,Female,73.8,62.0,72.0,65.0,95.0,48.0,3.0,2.0,0.0,7.16,Yes,2025-05-31
1021,1072,Pooja Chauhan,20.0,Female,87.5,72.0,52.0,80.0,81.0,63.0,4.0,2.0,2.0,7.67,Yes,2023-09-10
1022,1107,Rahul Singh,20.0,Female,76.8,unknown,76.0,76.0,96.0,72.0,4.0,3.0,1.0,7.46,No,2023-10-15
1023,1271,Manish Verma,21.0,NaN,83.8,64.0,29.0,69.0,89.0,-,6.0,1.0,1.0,6.97,Yes,2024-03-27


In [14]:
d_copy.drop_duplicates(inplace=True)
d_copy.duplicated().sum()

np.int64(0)

Student_ID      1025 non-null   int64  
 1   Name            1017 non-null   str    
 2   Age             1020 non-null   float64
 3   Gender          1010 non-null   str    
 4   Attendance      1014 non-null   str    
 5   Python          1016 non-null   str    
 6   DSA             1002 non-null   str    
 7   DBMS            1008 non-null   str    
 8   Maths           1011 non-null   str    
 9   OS              1003 non-null   str    
 10  Projects        1012 non-null   float64
 11  Internships     1011 non-null   float64
 12  Backlogs        1018 non-null   float64
 13  CGPA            1006 non-null   str    
 14  Placed          1013 non-null   str    
 15  Admission_Date  1006 non-null   str    

In [25]:
d_copy["Attendance"]=pd.to_numeric(d_copy["Attendance"], errors="coerce").astype("float64")
a=d_copy["Attendance"].mean()
d_copy["Attendance"]=d_copy["Attendance"].fillna(a)
d_copy["Attendance"].isnull().sum()

np.int64(0)

In [28]:
d_copy.info()

<class 'pandas.DataFrame'>
Index: 1000 entries, 0 to 1024
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Student_ID      1000 non-null   int64  
 1   Name            992 non-null    str    
 2   Age             995 non-null    float64
 3   Gender          985 non-null    str    
 4   Attendance      1000 non-null   float64
 5   Python          992 non-null    str    
 6   DSA             978 non-null    str    
 7   DBMS            986 non-null    str    
 8   Maths           987 non-null    str    
 9   OS              979 non-null    str    
 10  Projects        987 non-null    float64
 11  Internships     987 non-null    float64
 12  Backlogs        993 non-null    float64
 13  CGPA            981 non-null    str    
 14  Placed          988 non-null    str    
 15  Admission_Date  982 non-null    str    
dtypes: float64(5), int64(1), str(10)
memory usage: 132.8 KB
